tutorials/Tutorial17_GPL.py

```diff
@@ -0,0 +1,249 @@
+        t5_model = AutoModelForSeq2SeqLM.from_pretrained(t5_name).cuda()
+
+        batch_size = 32
+        queries_per_doc = 3
+
+        for start_idx in tqdm(range(0, len(corpus), batch_size)):
+            corpus_batch = corpus[start_idx : start_idx + batch_size]
+            enc_inp = t5_tokenizer(
+                corpus_batch, max_length=max_seq_length, truncation=True, padding=True, return_tensors="pt"
+            )
+
+            outputs = t5_model.generate(
+                input_ids=enc_inp["input_ids"].cuda(),
+                attention_mask=enc_inp["attention_mask"].cuda(),
+                max_length=64,
+                do_sample=True,
+                top_p=0.95,
+                num_return_sequences=queries_per_doc,
+            )
+
+            decoded_output = t5_tokenizer.batch_decode(outputs, skip_special_tokens=True)
+
+            for idx, query in enumerate(decoded_output):
+                corpus_id = int(idx / queries_per_doc)
+                query_doc_pairs.append({"question": query, "document": corpus_batch[corpus_id]})
+
+    print("Generated queries:", len(query_doc_pairs))
+
+    """# Use PseudoLabelGenerator to genenerate Retriever adaptation training data
+    
+    PseudoLabelGenerator#run will execute all three steps of the GPL [algorithm](https://github.com/UKPLab/gpl#how-does-gpl-work):
+     1. Question generation - optional step
+     2. Negative mining
+     3. Pseudo labeling (margin scoring)
+    
+    The output of the `PseudoLabelGenerator` is the training data we'll use to adapt our `EmbeddingRetriever`
+    
+    """
+
+    from haystack.nodes.question_generator import QuestionGenerator
+    from haystack.nodes.label_generator import PseudoLabelGenerator
+
+    use_question_generator = False
+
+    if use_question_generator:
+        questions_producer = QuestionGenerator(
+            model_name_or_path="doc2query/msmarco-t5-base-v1",
+            max_length=64,
+            split_length=128,
+            batch_size=32,
+            num_queries_per_doc=3,
+        )
+
+    else:
+        questions_producer = query_doc_pairs
+
+    # We can use either QuestionGenerator or already generated questions in PseudoLabelGenerator
+    psg = PseudoLabelGenerator(questions_producer, retriever, max_questions_per_document=10, batch_size=32, top_k=10)
+    output, pipe_id = psg.run(documents=document_store.get_all_documents())
+
+    """# Update the Retriever
+    
+    Now that we have the generated training data produced by `PseudoLabelGenerator` we'll update the `EmbeddingRetriever`. Let's take a peek at the training data
+    """
+
+    output["gpl_labels"][0]
+
+    len(output["gpl_labels"])
+
+    retriever.train(output["gpl_labels"])
+
+    """## Verify EmbeddingRetriever has been adapted; save it for future use
+    
+    We'll repeat our query and verify the Retriever is now more aware of the completely new concept such as COVID (i.e. ranks #1 in our query).
```

Comment:
Let's repeat our query to see if the Retriever learned about COVID and can now rank it as #1 among the answers.